GameServerLib/Logic/GameObjects/AttackableUnits/AttackableUnit.cs

```diff
@@ -114,31 +114,7 @@ public virtual void Die(AttackableUnit killer)
                     return;
                 }
 
-                var gold = _game.Map.MapGameScript.GetGoldFor(this);
-                if (gold <= 0)
-                {
-                    return;
-                }
-
-                cKiller.Stats.Gold += gold;
-                _game.PacketNotifier.NotifyAddGold(cKiller, this, gold);
-                cKiller.ChampStats.MinionsKilled += 1;
-                if (this.Team == TeamId.TEAM_NEUTRAL)
-                {
-                    cKiller.ChampStats.NeutralMinionsKilled += 1;
-                }
-
-                if (cKiller.KillDeathCounter < 0)
-                {
-                    cKiller.ChampionGoldFromMinions += gold;
-                    _logger.LogCoreInfo($"Adding gold form minions to reduce death spree: {cKiller.ChampionGoldFromMinions}");
-                }
-
-                if (cKiller.ChampionGoldFromMinions >= 50 && cKiller.KillDeathCounter < 0)
-                {
-                    cKiller.ChampionGoldFromMinions = 0;
-                    cKiller.KillDeathCounter += 1;
-                }
+                cKiller.OnKillMinion(this as Minion);
```

Comment:
This can be dangerous as ``this as Minion`` may be null, therefore it may throw an NPE when you kill any ``AttackableUnit`` that is not a minion
I'd say, just do something like 
```cs
if (this is Minion) {
    cKiller.OnKillMinion((Minion) this);
}

Reply:
Also this OnKillMinion code should go to ``Minion.Die`` probably

Reply:
but OnKillMinion changes the state of the champion, if we want that hero stats will be private to set that is needed that this function will be in champion class, also I think that is bad to call ```OnKillMinion``` from ```Minion.Die```, and it is better to call it from where we call ```Minion.Die```.

I'll add a check for the type but in champion function, and will change the signature it to ```OnKill (AtttackableUnit killed)```.